src/Reflection/ReflectionClass.php

```diff
@@ -1588,39 +1630,52 @@ private function getCurrentClassImplementedInterfacesIndexedByName(): array
     /**
      * This method allows us to retrieve all interfaces parent of this interface. Do not use on class nodes!
      *
-     * @param list<class-string> $interfaceClassNames
+     * @param list<class-string> $classNameStack
      *
      * @return array<class-string, ReflectionClass> parent interfaces of this interface
      *
      * @throws NotAnInterfaceReflection
      */
-    private function getInterfacesHierarchy(array &$interfaceClassNames = []): array
+    private function getInterfacesHierarchy(array &$classNameStack = []): array
     {
         if (! $this->isInterface) {
             throw NotAnInterfaceReflection::fromReflectionClass($this);
         }
 
         $interfaceClassName = $this->getName();
-        if (in_array($interfaceClassName, $interfaceClassNames, true)) {
-            throw CircularReference::fromClassName($interfaceClassName);
-        }
-
-        $interfaceClassNames[] = $interfaceClassName;
+        $classNameStack     = $this->pushClassNameStack($classNameStack, $interfaceClassName);
 
         /** @var array<class-string, self> $interfaces */
         $interfaces = array_merge(
             [$interfaceClassName => $this],
             ...array_map(
                 fn (string $interfaceClassName): array => $this->reflector
                     ->reflectClass($interfaceClassName)
-                    ->getInterfacesHierarchy($interfaceClassNames),
+                    ->getInterfacesHierarchy($classNameStack),
                 $this->implementsClassNames,
             ),
         );
 
         return $this->addStringableInterface($interfaces);
     }
 
+    /**
+     * @param list<class-string> $classNameStack
+     * @param class-string       $className
+     *
+     * @return list<class-string>
+     */
+    private function pushClassNameStack(array $classNameStack, string $className): array
+    {
+        if (in_array($className, $classNameStack, true)) {
+            throw CircularReference::fromClassName($className);
+        }
+
+        $classNameStack[] = $className;
+
+        return $classNameStack;
+    }
```

Comment:
Minor improvement: can we make this `array<class-string, null>` and prevent usage of `in_array()` (which does a linear scan)?